### PR TITLE
feat(BUX-149): fee quotes implemented

### PR DESCRIPTION
**Pull request by kuba-4chain**

### CODE REVIEW

File: broadcast/internal/composite/broadcaster.go [hunk 1]
```diff
@@ -31,6 +34,120 @@ func NewBroadcaster(strategy Strategy, factories ...BroadcastFactory) broadcast.
...
+
+    feeQuote, ok := result.(*broadcast.FeeQuote)
+    if !ok {
+        return nil, fmt.Errorf("unexpected result type: %T", result)
+    }
+
+    return feeQuote, nil
+}
+
+func (c *compositeBroadcaster) GetBestQuote(ctx context.Context) (*broadcast.FeeQuote, error) {
```
Comment (arkadiuszos4chain) [inline broadcast/internal/composite/broadcaster.go:79]: If we want to query all broadcaster one by one I would prefer alloc-free version like:

```go
var bestQuote *broadcast.FeeQuote = nil

for _, broadcaster := range c.broadcasters {
	feeQuote, err := broadcaster.GetFeeQuote(ctx)
	if err == nil {

		if bestQuote == nil {
			bestQuote = feeQuote
		} else {
			feePer1000Bytes := (1000 / feeQuote.MiningFee.Bytes) * feeQuote.MiningFee.Satoshis
			bestFeePer1000Bytes := (1000 / bestQuote.MiningFee.Bytes) * bestQuote.MiningFee.Satoshis
			
			if feePer1000Bytes < bestFeePer1000Bytes {
				bestQuote = feeQuote
			}
		}

	}
}


if bestQuote == nil {
	return nil, errors.New("failed to get reponse from any miner")
}

return best, nil

```

TBH I would prefer querying broadcasters in parallel

File: broadcast/internal/composite/broadcaster.go [hunk 2]
```diff
@@ -31,6 +34,120 @@ func NewBroadcaster(strategy Strategy, factories ...BroadcastFactory) broadcast.
...
+            }
+        } (ctxWithTimeout, broadcaster)
+    }
+
+    go func() {
+        wg.Wait()
+        close(feeChannel)
+    }()
+
+    fastQuote := <-feeChannel
```
Comment (arkadiuszos4chain) [inline broadcast/internal/composite/broadcaster.go:142]: What if nothing ever hits the channel? ;-)

File: broadcast/internal/composite/broadcaster.go [hunk 3]
```diff
@@ -31,6 +34,120 @@ func NewBroadcaster(strategy Strategy, factories ...BroadcastFactory) broadcast.
...
+
+    if fastQuote == nil {
```
Comment (arkadiuszos4chain) [inline broadcast/internal/composite/broadcaster.go:144]: As I understand it, we should wait for the fastest correct quote, so we should make sure `nil` never enters the channel

**Review by arkadiuszos4chain — CHANGES_REQUESTED** [review]

File: broadcast/internal/arc/arc_policy_quote.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	url := arc.apiURL + arcPolicyQuoteRoute
+	pld := httpclient.NewPayload(
+		httpclient.GET,
+		url,
+		arc.token,
+		nil,
+	)
+
+	resp, err := arc.HTTPClient.DoRequest(ctx, pld)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply broadcast/internal/arc/arc_policy_quote.go:36]: Alright, done

File: broadcast/internal/arc/arc_policy_quote.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	model, err := decodeQueryPolicyBody(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return model, nil
+}
+
+func decodeQueryPolicyBody(body io.ReadCloser) (*broadcast.PolicyQuoteResponse, error) {
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply broadcast/internal/arc/arc_policy_quote.go]: Good point, done :)

File: broadcast/internal/composite/broadcaster.go (hunk 2 above)
Comment (kuba-4chain) [reply broadcast/internal/composite/broadcaster.go:145]: Yes, exactly :)

File: broadcast/internal/composite/broadcaster.go (hunk 3 above)
Comment (kuba-4chain) [reply broadcast/internal/composite/broadcaster.go:147]: Nil will enter the channel when all miners will timeout or throw an error (or some miners will timeout and the rest will throw an error), so basically when there is zero responses from all the queries.

File: broadcast/internal/composite/broadcaster.go (hunk 1 above)
Comment (arkadiuszos4chain) [reply broadcast/internal/composite/broadcaster.go:79]: Nice ;-)

File: broadcast/internal/composite/broadcaster.go (hunk 3 above)
Comment (arkadiuszos4chain) [reply broadcast/internal/composite/broadcaster.go:147]: Yes, you're right. My bad ;-)

**Review by arkadiuszos4chain — APPROVED** [review]

**Pull request merged**